packages/synapse-core/src/piece.ts

```diff
@@ -89,3 +262,234 @@ export function calculate(data: Uint8Array) {
+): Promise<Uint8Array> {
+  // Parse and validate the expected PieceCID
+  const parsedPieceCid = asPieceCID(expectedPieceCid)
+  if (parsedPieceCid == null) {
+    throw new Error(`Invalid PieceCID: ${String(expectedPieceCid)}`)
+  }
+
+  // Check response is OK
+  if (!response.ok) {
+    throw new Error(`Download failed: ${response.status} ${response.statusText}`)
+  }
+
+  if (response.body == null) {
+    throw new Error('Response body is null')
+  }
+
+  // Create PieceCID calculation stream
+  const { stream: pieceCidStream, getPieceCID } = createPieceCIDStream()
+
+  // Create a stream that collects all chunks into an array
+  const chunks: Uint8Array[] = []
+  const collectStream = new TransformStream<Uint8Array, Uint8Array>({
+    transform(chunk: Uint8Array, controller: TransformStreamDefaultController<Uint8Array>) {
+      chunks.push(chunk)
+      controller.enqueue(chunk)
+    },
+  })
+
+  // Pipe the response through both streams
+  const pipelineStream = response.body.pipeThrough(pieceCidStream).pipeThrough(collectStream)
+
+  // Consume the stream to completion
+  const reader = pipelineStream.getReader()
+  try {
+    while (true) {
+      const { done } = await reader.read()
+      if (done) break
+    }
+  } finally {
+    reader.releaseLock()
+  }
+
+  if (chunks.length === 0) {
+    throw new Error('Response body is empty')
+  }
+
+  // Get the calculated PieceCID
+  const calculatedPieceCid = getPieceCID()
+
+  if (calculatedPieceCid == null) {
+    throw new Error('Failed to calculate PieceCID from stream')
+  }
+
+  // Verify the PieceCID
+  if (calculatedPieceCid.toString() !== parsedPieceCid.toString()) {
+    throw new Error(
+      `PieceCID verification failed. Expected: ${String(parsedPieceCid)}, Got: ${String(calculatedPieceCid)}`
+    )
+  }
+
+  // Combine all chunks into a single Uint8Array
+  const totalLength = chunks.reduce((acc, chunk) => acc + chunk.length, 0)
+  const result = new Uint8Array(totalLength)
+  let offset = 0
+  for (const chunk of chunks) {
+    result.set(chunk, offset)
+    offset += chunk.length
+  }
+
+  return result
+}
+
+/**
+ * Download data from a URL, validate its PieceCID, and return as Uint8Array
+ *
+ * This is a convenience function that fetches from a URL and then uses
+ * downloadAndValidate to download and validate the data.
+ *
+ * @param url - The URL to download from
+ * @param expectedPieceCid - The expected PieceCID to validate against
+ * @returns The downloaded data as a Uint8Array
+ * @throws Error if PieceCID validation fails or download errors occur
+ *
+ * @example
+ * ```typescript
+ * const data = await downloadAndValidateFromUrl(
+ *   'https://provider.com/piece/bafkzcib...',
+ *   'bafkzcib...'
+ * )
+ * ```
+ */
+export async function downloadAndValidateFromUrl(
+  url: string,
+  expectedPieceCid: string | PieceCID
+): Promise<Uint8Array> {
+  const response = await fetch(url)
+  return await downloadAndValidate(response, expectedPieceCid)
+}
```

Comment:
I feel like downloadAndValidate is nice enough to where we probably don't need this. If we do.. will we ever need to override fetch options? should we accept InitOptions as well as URL | string?